### PR TITLE
set 100000 fds on systemd unit creation if unspecified

### PR DESCRIPTION
**Pull request by knisbet**
## Description

Some distros / versions of systemd default to 1024 open fds. So by default if unspecified, create systemd units from gravity with 100000 fds as 1024 is far too low for gravity daemons by default.

## Type of change

* Bug fix (non-breaking change which fixes an issue)

## Linked tickets and other PRs

Updates #2314

## TODOs

- [x] Self-review the change
- [x] Write tests
- [ ] Address review feedback

## Implementation

I just set this to infinity if not specified, to avoid any defaults / limits. 

## Performance/Scaling

I suppose in theory this could cause issues with lots of open connections that will present in a different way than being blocked due to max open fds.

## Testing done

Created an equivelant change and tested it on a test cluster. New Unit tests. Relying on robotest for this.

### CODE REVIEW

**Review by lenko-d — COMMENTED** [review]
I think that we should make it the same as the current value in Planet. This is for the sake of consistency and also security. 

From a security point of view having a resource that is unlimited means that an attacker could potentially bring the system to its knees by opening too many fds. Having some resource limits is the simplest form of protection.

The value in Planet should be big enough:

```
cat /proc/sys/fs/file-max
1619372

```

**Comment by knisbet** [comment]
> I think that we should make it the same as the current value in Planet. This is for the sake of consistency and also security.
> 
> From a security point of view having a resource that is unlimited means that an attacker could potentially bring the system to its knees by opening too many fds. Having some resource limits is the simplest form of protection.
> 
> I think that we should put the same value as in Planet. It should be big enough:
> 
> ```
> cat /proc/sys/fs/file-max
> 1619372
> ```

Setting to unlimited in the systemd configuration on LimitNoFile doesn't actually make the fds unlimited in the kernel.
```
root@kevin-test1:~# cat /proc/28039/limits
Limit                     Soft Limit           Hard Limit           Units
Max cpu time              unlimited            unlimited            seconds
Max file size             unlimited            unlimited            bytes
Max data size             unlimited            unlimited            bytes
Max stack size            8388608              unlimited            bytes
Max core file size        0                    unlimited            bytes
Max resident set          unlimited            unlimited            bytes
Max processes             60041                60041                processes
Max open files            1048576              1048576              files
Max locked memory         65536                65536                bytes
Max address space         unlimited            unlimited            bytes
Max file locks            unlimited            unlimited            locks
Max pending signals       60041                60041                signals
Max msgqueue size         819200               819200               bytes
Max nice priority         0                    0
Max realtime priority     0                    0
Max realtime timeout      unlimited            unlimited            us
```

This is not actually file-max, so I don't know where it pulls from.

**Comment by lenko-d** [comment]
> Setting to unlimited in the systemd configuration on LimitNoFile doesn't actually make the fds unlimited in the kernel.

What is more important is that you are declaring an intention to have a resource that is unlimited. It may not end up being unlimited in reality but the fact that you define it like that doesn't sound good from a security point of view.

I'm not saying that Gravity will not pass a security audit because it has defined unlimited resource that could possibly enable denial of service attacks. I'm just saying that it doesn't look good to me to declare a resource as unlimited.

**Comment by knisbet** [comment]
> > Setting to unlimited in the systemd configuration on LimitNoFile doesn't actually make the fds unlimited in the kernel.
> 
> What is more important is that you are declaring an intention to have a resource that is unlimited. It may not end up being unlimited in reality but the fact that you define it like that doesn't sound good from a security point of view.
> 
> I'm not saying that Gravity will not pass a security audit because it has defined unlimited resource that could possibly enable denial of service attacks. I'm just saying that it doesn't look good to me to declare a resource as unlimited.


That's fair, sometimes perceived problems are worth addressing, I changed it to 100k which should only be hit if there is a legit problem.

**Review by lenko-d — APPROVED** [review]

File: lib/systemservice/systemd_test.go [hunk 1]
```diff
@@ -120,7 +125,91 @@ TasksMax=infinity
 [Install]
 WantedBy=test.target
 
-`)
+`,
+		},
+		{
+			description: "unspecified file limits set to infinity",
```
Comment (lenko-d) [inline lib/systemservice/systemd_test.go:131]: Infinity is not exactly correct.

**Review by lenko-d — CHANGES_REQUESTED** [review]

**Review by lenko-d — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

File: lib/systemservice/systemd_test.go [hunk 2]
```diff
@@ -120,7 +125,91 @@ TasksMax=infinity
...
+WantedBy=test.target
+
+`,
+		},
+	}
+
+	for _, test := range tt {
+		buf := &bytes.Buffer{}
+		err := serviceUnitTemplate.Execute(buf, test.in)
+		c.Assert(err, IsNil)
```
Comment (bernardjkim) [inline lib/systemservice/systemd_test.go:210]: nit: Include comment

**Review by bernardjkim — APPROVED** [review]

**Pull request merged**